### PR TITLE
filter by requester

### PR DESCRIPTION
**Pull request by se3000**
By allowing an oracle node to filter by requester, job spec creators can permission specific addresses.

### CODE REVIEW

File: services/subscription.go [hunk 1]
```diff
@@ -153,7 +156,7 @@ func TopicFiltersForRunLog(jobID string) [][]common.Hash {
 	hexJobID := common.BytesToHash([]byte(jobID))
 	jobIDZeroPadded := common.BytesToHash(common.RightPadBytes(hexutil.MustDecode("0x"+jobID), utils.EVMWordByteLen))
```
Comment (j16r) [inline services/subscription.go]: `jobIDZeroPadded` is a little bit of a weird name for a `Hash`, because I think of zero padding strings, but not hashes (which are byte arrays).

File: services/subscription.go [hunk 2]
```diff
@@ -409,11 +427,23 @@ func (le InitiatorSubscriptionLogEvent) ContractPayment() (*assets.Link, error)
 	return payment, nil
 }
 
+// Requester pulls the requesting address out of the LogEvent's topics.
+func (le InitiatorSubscriptionLogEvent) Requester() common.Address {
+	b := le.Log.Topics[RunLogTopicRequester].Bytes()
+	return common.BytesToAddress(b)
+}
+
+func parseRequestID(data hexutil.Bytes) string {
```
Comment (j16r) [inline services/subscription.go]: `parsing` is really for strings, seems this is doing more of an `encoding` job?

File: services/subscription.go [hunk 3]
```diff
@@ -409,11 +427,23 @@ func (le InitiatorSubscriptionLogEvent) ContractPayment() (*assets.Link, error)
 	return payment, nil
 }
 
+// Requester pulls the requesting address out of the LogEvent's topics.
+func (le InitiatorSubscriptionLogEvent) Requester() common.Address {
+	b := le.Log.Topics[RunLogTopicRequester].Bytes()
+	return common.BytesToAddress(b)
```
Comment (j16r) [inline services/subscription.go]: Hmm, does this mean the address is being stores in the first 20 bytes, then the remaining 12 bytes are 0s? I was under the impression the 0s are the higher endian bytes.

File: services/subscription.go [hunk 4]
```diff
@@ -22,16 +23,18 @@ import (
 // Descriptive indices of a RunLog's Topic array
 const (
 	RunLogTopicSignature = iota
-	RunLogTopicInternalID
 	RunLogTopicJobID
+	RunLogTopicRequester
 	RunLogTopicAmount
 )
 
+const evmWordBytesSize = 32
```
Comment (j16r) [inline services/subscription.go]: I think this constant is also in `go-ethereum` as ... `HashLength` ?

File: services/subscription.go [hunk 5]
```diff
@@ -409,11 +427,23 @@ func (le InitiatorSubscriptionLogEvent) ContractPayment() (*assets.Link, error)
...
 }
 
+// Requester pulls the requesting address out of the LogEvent's topics.
+func (le InitiatorSubscriptionLogEvent) Requester() common.Address {
+	b := le.Log.Topics[RunLogTopicRequester].Bytes()
+	return common.BytesToAddress(b)
+}
+
+func parseRequestID(data hexutil.Bytes) string {
+	parsedID := []byte(data)[:evmWordBytesSize]
```
Comment (j16r) [inline services/subscription.go]: Rather than slicing here, could we accept a fixed size input like `common.Hash`?

File: internal/cltest/factories.go [hunk 6]
```diff
@@ -270,14 +277,14 @@ func StringToVersionedLogData(str string) hexutil.Bytes {
...
-	offset := "0000000000000000000000000000000000000000000000000000000000000020"
+	version := utils.RemoveHexPrefix(utils.EVMHexNumber(1))
+	offset := "0000000000000000000000000000000000000000000000000000000000000060"
 
 	var endPad string
 	if length%32 != 0 {
 		endPad = strings.Repeat("00", (32 - (length % 32)))
 	}
-	return hexutil.MustDecode(version + offset + lenHex + data + endPad)
+	return hexutil.MustDecode(id + version + offset + lenHex + data + endPad)
```
Comment (j16r) [inline internal/cltest/factories.go]: This may be because of my systems programming/C background, but I generally prefer to do things bytewise rather than string wise, what do you say to refactoring this a little bit?

**Comment by j16r** [comment]
This is good, but I think there's a bit of a mixing `string`s and `[]byte`s we're doing in the code base which might be worth cleaning up a bit here.

File: services/subscription.go (hunk 1 above)
Comment (se3000) [reply services/subscription.go]: I agree. I was following the naming convention of the go-ethereum codebase, where as far as I can tell `common.Hash` is used to refer to anything that is 32 bytes long. For example, [common.BytesToHash](https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L51) doesn't hash anything, it just truncates the number of bytes and returns them in an array. I would normally refer to that as an "EVM word." Maybe that's a better name?

File: services/subscription.go (hunk 3 above)
Comment (se3000) [reply services/subscription.go]: You're right that the higher order bytes the zeroes. [BytesToAddress takes the lower endian bytes.](https://github.com/ethereum/go-ethereum/blob/master/common/types.go#L231)

File: internal/cltest/factories.go (hunk 6 above)
Comment (se3000) [reply internal/cltest/factories.go]: I think that setting things like `offset` would be significantly less straightforward if done in binary. This is also a test helper, so I tend to prioritize readability over performance.

File: services/subscription.go (hunk 1 above)
Comment (j16r) [reply services/subscription.go]: https://solidity.readthedocs.io/en/develop/abi-spec.html explains it a little more: the EVM stores variable sized values (such as `bytes`) in the higher order bits of a word, with big endian encoding.

File: internal/cltest/factories.go [hunk 7]
```diff
@@ -265,26 +265,18 @@ func NewRunLog(
...
-	j := JSONFromString(str)
-	cbor, err := j.CBOR()
-	mustNotErr(err)
-	length := len(cbor)
-	lenHex := utils.RemoveHexPrefix(hexutil.EncodeUint64(uint64(length)))
-	if len(lenHex) < 64 {
-		lenHex = strings.Repeat("0", 64-len(lenHex)) + lenHex
-	}
+func StringToVersionedLogData(internalID, str string) []byte {
+	buf := bytes.NewBuffer(hexutil.MustDecode(StringToHash(internalID).Hex()))
```
Comment (j16r) [inline internal/cltest/factories.go]: I believe you can shorten this to `buf := bytes.NewBuffer(StringToHash(internalID).Bytes())`
